transaction.go

```diff
@@ -162,313 +191,512 @@ func (t *Transaction) UnmarshalJSON(bytes []byte) error {
-			return 0, nil, fmt.Errorf("%w: UTXO for ID %v is not provided (input at index %d)", ErrMissingUTXO, utxoID, i)
-		}
+// TxSemanticDeposit validates that the IOTA tokens are balanced from the input/output side.
+// It additionally also incorporates the check whether return amounts via StorageDepositReturnUnlockCondition(s) for specified identities
+// are fulfilled from the output side.
+func TxSemanticDeposit() TxSemanticValidationFunc {
+	return func(svCtx *SemanticValidationContext) error {
+		// note that due to syntactic validation of outputs, input and output deposit sums
+		// are always within bounds of the total token supply
+		var in, out uint64
+		inputSumReturnAmountPerIdent := make(map[string]uint64)
+		for inputID, input := range svCtx.WorkingSet.InputSet {
+			in += input.Deposit()
+
+			unlockCondSet := input.UnlockConditions().MustSet()
+			storageDepositReturnUnlockCondition := unlockCondSet.StorageDepositReturn()
+			if storageDepositReturnUnlockCondition == nil {
+				continue
+			}
 
-		var err error
-		deposit, err := utxo.Deposit()
-		if err != nil {
-			return 0, nil, fmt.Errorf("unable to get deposit from UTXO %v (input at index %d): %w", utxoID, i, err)
-		}
-		inputSum += deposit
+			if !unlockCondSet.HasExpirationCondition() {
+				continue
+			}
 
-		sigBlock, sigBlockIndex, err := t.signatureUnlockBlock(i)
-		if err != nil {
-			return 0, nil, err
-		}
+			storageDepositReturnIdentKey := storageDepositReturnUnlockCondition.ReturnAddress.Key()
 
-		target, err := utxo.Target()
-		if err != nil {
-			return 0, nil, fmt.Errorf("unable to get target for UTXO %v: %w", utxoID, err)
-		}
+			// if the return ident unlocked this input, then the return amount does
+			// not have to be fulfilled
+			unlockedIndices, has := svCtx.WorkingSet.UnlockedIdents[storageDepositReturnIdentKey]
+			if has && unlockedIndices.Unlocked(svCtx.WorkingSet.InputIDToIndex[inputID]) {
+				continue
+			}
 
-		// change this logic here once we got tx output types without addrs
-		addr, isAddr := target.(Address)
-		if !isAddr {
-			return 0, nil, fmt.Errorf("target for UTXO %v must be an address: %w", utxoID, err)
+			inputSumReturnAmountPerIdent[storageDepositReturnIdentKey] += storageDepositReturnUnlockCondition.Amount
 		}
 
-		usedSigBlockIndex, alreadySeen := seenInputAddr[addr.String()]
-		if alreadySeen {
-			if usedSigBlockIndex != sigBlockIndex {
-				return 0, nil, fmt.Errorf("%w: target for UTXO %v uses a different signature unlock block (%d) than a previous UTXO (%d) for the same address", ErrInputSignatureUnlockBlockInvalid, utxoID, sigBlockIndex, usedSigBlockIndex)
+		outputSimpleTransfersPerIdent := make(map[string]uint64)
+		for _, output := range svCtx.WorkingSet.Tx.Essence.Outputs {
+			outDeposit := output.Deposit()
+			out += outDeposit
+
+			// accumulate simple transfers for StorageDepositReturnUnlockCondition checks
+			if basicOutput, is := output.(*BasicOutput); is {
+				if len(basicOutput.FeatureBlocks()) > 0 {
```

Comment:
This must have been forgotten here since we introduced unlock conditions instead of feature blocks.
The check should rather be:
 - The basic output only has one unlock condition, and that is Address Unlock.